exp/zapslog/slog.go

```diff
@@ -0,0 +1,145 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package zapslog
+
+import (
+	"context"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+	"golang.org/x/exp/slog"
+)
+
+// Handler implements the slog.Handler by writing to a zap Core.
+type Handler struct {
+	core zapcore.Core
+}
+
+var _ slog.Handler = (*Handler)(nil)
+
+// groupObject holds all the Attrs saved in a slog.GroupValue.
+type groupObject []slog.Attr
+
+func (gs groupObject) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	for _, attr := range gs {
+		convertAttrToField(attr).AddTo(enc)
+	}
+	return nil
+}
+
+func convertAttrToField(attr slog.Attr) zapcore.Field {
+	switch attr.Value.Kind() {
+	case slog.KindBool:
+		return zap.Bool(attr.Key, attr.Value.Bool())
+	case slog.KindDuration:
+		return zap.Duration(attr.Key, attr.Value.Duration())
+	case slog.KindFloat64:
+		return zap.Float64(attr.Key, attr.Value.Float64())
+	case slog.KindInt64:
+		return zap.Int64(attr.Key, attr.Value.Int64())
+	case slog.KindString:
+		return zap.String(attr.Key, attr.Value.String())
+	case slog.KindTime:
+		return zap.Time(attr.Key, attr.Value.Time())
+	case slog.KindUint64:
+		return zap.Uint64(attr.Key, attr.Value.Uint64())
+	case slog.KindGroup:
+		return zap.Object(attr.Key, groupObject(attr.Value.Group()))
+	case slog.KindLogValuer:
+		return convertAttrToField(slog.Attr{
+			Key: attr.Key,
+			// FIXME(knight42): resolve the value in a lazy way
+			Value: attr.Value.Resolve(),
+		})
+	default:
+		return zap.Any(attr.Key, attr.Value.Any())
+	}
+}
+
+// convertSlogLevel maps slog Levels to zap Levels.
+// Note that there is some room between slog levels while zap levels are continuous, so we can't 1:1 map them.
+// See also https://go.googlesource.com/proposal/+/master/design/56345-structured-logging.md?pli=1#levels
+func convertSlogLevel(l slog.Level) zapcore.Level {
+	switch {
+	case l >= slog.LevelError:
+		return zapcore.ErrorLevel
+	case l >= slog.LevelWarn:
+		return zapcore.WarnLevel
+	case l >= slog.LevelInfo:
+		return zapcore.InfoLevel
+	default:
+		return zapcore.DebugLevel
+	}
+}
+
+func (h *Handler) Enabled(ctx context.Context, level slog.Level) bool {
+	return h.core.Enabled(convertSlogLevel(level))
+}
+
+func (h *Handler) Handle(ctx context.Context, record slog.Record) error {
+	ent := zapcore.Entry{
+		Level:   convertSlogLevel(record.Level),
+		Time:    record.Time,
+		Message: record.Message,
+		// FIXME: do we need to set the following fields?
+		// LoggerName:
+		// Caller:
+		// Stack:
+	}
```

Comment:
cc @abhinav . Do we need to complete the `Entry`?

Reply:
Those three fields are optional. We can start without them.
I can see Caller and LoggerName being filled in the future.
(We could probably fill the Caller from [Record.PC](https://pkg.go.dev/golang.org/x/exp/slog#Record.PC),
but doing that cheaply adds some complexity so we can do without for now.)

However, this does bring up the need for customization.
A plain `NewHandler` may not be sufficient for that.

I've pushed a change to your branch with the following new API:

```go
type HandlerOptions struct{ LoggerName string }

func (HandlerOptions) New(zapcore.Core) *Handler
```

The `func NewHandler(..)` just calls this now to get default options.
This mirrors the JSONHandler, NewJSONHandler, HandlerOptions setup.

This leaves us room for future customization llke:
opting into adding a caller name, custom level mapping, etc.

Reply:
Thank you! The `HandlerOptions` is commonly used for customization, though I thought we might want to consistently follow the functional options pattern in zap.

> I can see Caller and LoggerName being filled in the future.
(We could probably fill the Caller from [Record.PC](https://pkg.go.dev/golang.org/x/exp/slog#Record.PC),

After some trial & error, I found I could add the Caller in this PR as well, plz see the commit https://github.com/uber-go/zap/pull/1246/commits/1ae59453aac0c7f3e7fec679ca167393f7b47218